src/nvim/strings.c

```diff
@@ -335,36 +335,40 @@ void vim_strup(char *p)
 char *strcase_save(const char *const orig, bool upper)
   FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
 {
-  char *res = xstrdup(orig);
+  // Calculate the initial length and allocate memory for the result
+  size_t orig_len = strlen(orig);
+  // +1 for the null terminator
+  char *res = xmalloc(orig_len + 1);
+  // Index in the result string
+  size_t res_index = 0;
+  // Current position in the original string
+  const char *p = orig;
 
-  char *p = res;
   while (*p != NUL) {
-    int c = utf_ptr2char(p);
-    int l = utf_ptr2len(p);
-    if (c == 0) {
-      // overlong sequence, use only the first byte
-      c = (uint8_t)(*p);
-      l = 1;
-    }
-    int uc = upper ? mb_toupper(c) : mb_tolower(c);
-
-    // Reallocate string when byte count changes.  This is rare,
-    // thus it's OK to do another malloc()/free().
-    int newl = utf_char2len(uc);
-    if (newl != l) {
-      // TODO(philix): use xrealloc() in strcase_save()
-      char *s = xmalloc(strlen(res) + (size_t)(1 + newl - l));
-      memcpy(s, res, (size_t)(p - res));
-      STRCPY(s + (p - res) + newl, p + l);
-      p = s + (p - res);
-      xfree(res);
-      res = s;
-    }
-
-    utf_char2bytes(uc, p);
-    p += newl;
-  }
-
+    CharInfo char_info = utf_ptr2CharInfo(p);
+    int uc = upper ? mb_toupper(char_info.value) : mb_tolower(char_info.value);
+    // Cast to size_t for the same reason
```

Comment:
The reason is now missing
```suggestion
    // Cast to size_t to avoid mixing types in arithmetic
```